server/src/main/java/org/apache/lucene/index/BaseStarTreeBuilder.java

```diff
@@ -0,0 +1,647 @@
+        this.fieldProducerMap = fieldProducerMap;
+        this.starTreeDocValuesIteratorAdapter = new StarTreeDocValuesIteratorAdapter();
+
+        List<Dimension> dimensionsSplitOrder = starTreeField.getDimensionsOrder();
+        this.numDimensions = dimensionsSplitOrder.size();
+
+        this.skipStarNodeCreationForDimensions = new HashSet<>();
+        this.totalSegmentDocs = state.segmentInfo.maxDoc();
+        this.dimensionReaders = new SequentialDocValuesIterator[numDimensions];
+        Set<String> skipStarNodeCreationForDimensions = starTreeFieldSpec.getSkipStarNodeCreationInDims();
+
+        for (int i = 0; i < numDimensions; i++) {
+            String dimension = dimensionsSplitOrder.get(i).getField();
+            if (skipStarNodeCreationForDimensions.contains(dimensionsSplitOrder.get(i).getField())) {
+                this.skipStarNodeCreationForDimensions.add(i);
+            }
+            FieldInfo dimensionFieldInfos = state.fieldInfos.fieldInfo(dimension);
+            DocValuesType dimensionDocValuesType = dimensionFieldInfos.getDocValuesType();
+            dimensionReaders[i] = starTreeDocValuesIteratorAdapter.getDocValuesIterator(
+                dimensionDocValuesType,
+                dimensionFieldInfos,
+                fieldProducerMap.get(dimensionFieldInfos.name)
+            );
+        }
+
+        this.metricAggregatorInfos = generateMetricAggregatorInfos(mapperService, state);
+        this.numMetrics = metricAggregatorInfos.size();
+        this.maxLeafDocuments = starTreeFieldSpec.maxLeafDocs();
+    }
+
+    /**
+     * Generates the configuration required to perform aggregation for all the metrics on a field
+     *
+     * @return list of MetricAggregatorInfo
+     */
+    public List<MetricAggregatorInfo> generateMetricAggregatorInfos(MapperService mapperService, SegmentWriteState state)
+        throws IOException {
+        List<MetricAggregatorInfo> metricAggregatorInfos = new ArrayList<>();
+        for (Metric metric : this.starTreeField.getMetrics()) {
+            for (MetricStat metricStat : metric.getMetrics()) {
+                IndexNumericFieldData.NumericType numericType;
+                SequentialDocValuesIterator metricStatReader;
+                Mapper fieldMapper = mapperService.documentMapper().mappers().getMapper(metric.getField());
+                if (fieldMapper instanceof NumberFieldMapper) {
+                    numericType = ((NumberFieldMapper) fieldMapper).fieldType().numericType();
+                } else {
+                    logger.error("unsupported mapper type");
+                    throw new IllegalStateException("unsupported mapper type");
+                }
+
+                FieldInfo metricFieldInfos = state.fieldInfos.fieldInfo(metric.getField());
+                DocValuesType metricDocValuesType = metricFieldInfos.getDocValuesType();
+                if (metricStat != MetricStat.COUNT) {
```

Comment:
in case of Count, we don't depend on the underlying metric values instead perform a count on top of dimension values. In this case, we need not initialize an actual metric reader for this metric.